### PR TITLE
front: migration of bootstrap component to typescript

### LINKED ISSUE #4960: front: migrate common/bootstrapSNCF files to typescript

### PR DESCRIPTION
**Pull request by sim51**
closes #4960

List of unused component that have been deleted : 

* ListGroupSNCF
* AlertSNCF
* FloatingBar
* MastNavItemSNCF
* MastNavSNCF
* PaginationSNCF
* ProgressSNCF
* SideBarSNCF

Two components have been hard to migrate : the select components. 

I did some modification on them : 
* rename `selectedvalue` attribut to `value` like usual input
* rename `title` to `label`, title can be interpreted like the title of the input 

`SelectImprovedSNCF` can only takes a list of `string` or a list of `{id: string, label:string}`.  

NB: IMHO, the `SelectImprovedSNCF` should be fully replaced by a `react-select`, which is a standard

### CODE REVIEW

File: front/src/common/BootstrapSNCF/CardSNCF/CardSNCF.tsx
Comment (clarani) [inline front/src/common/BootstrapSNCF/CardSNCF/CardSNCF.tsx]:
```suggestion
  <Link to={link} className={`card overflow-hidden mb-2 ${disabledLink && 'disabled-link'}`}>
```

File: front/src/common/BootstrapSNCF/CheckboxRadioSNCF.tsx
Comment (clarani) [inline front/src/common/BootstrapSNCF/CheckboxRadioSNCF.tsx]:
```suggestion
/*
```

File: front/src/common/BootstrapSNCF/SelectSNCF.tsx
Comment (clarani) [inline front/src/common/BootstrapSNCF/SelectSNCF.tsx]: It's only a suggestion, I find this clearer, but I'll let you decide what you prefer. You can resolve this comment on your own.

```suggestion
        {options.map((option) => {
          const optionId = isString(option) ? option : option.id;
          const optionLabel = isString(option) ? option : option.label;
          return (
            <option key={optionId} value={optionId}>
              {optionLabel}
            </option>
          );
        })}
```

File: front/src/common/BootstrapSNCF/SelectImprovedSNCF.tsx
Comment (clarani) [inline front/src/common/BootstrapSNCF/SelectImprovedSNCF.tsx]: No need to set default value to false
```suggestion
  inline,
  label,
  options,
  value,
  onChange,
  sm,
  withSearch,
  withNewValueInput,
  addButtonTitle = 'Add',
  bgWhite,
  dataTestId = '',
```

File: front/src/common/BootstrapSNCF/SelectImprovedSNCF.tsx
Comment (clarani) [inline front/src/common/BootstrapSNCF/SelectImprovedSNCF.tsx]: Maybe we can use a useMemo here

File: front/src/common/BootstrapSNCF/SelectImprovedSNCF.tsx
Comment (clarani) [inline front/src/common/BootstrapSNCF/SelectImprovedSNCF.tsx]:
```suggestion
  const optionsComponents = useMemo(() =>
    filteredOptions.map((option) => (
      <span className="select-menu-item" role="listitem" key={nextId()}>
        <button type="button" onClick={() => selectItem(option)}>
          {typeof option === 'string' ? option : option.label}
        </button>
      </span>
    )), [filteredOptions]);
```

File: front/src/common/BootstrapSNCF/SelectImprovedSNCF.tsx
Comment (clarani) [inline front/src/common/BootstrapSNCF/SelectImprovedSNCF.tsx]:
```suggestion
```

**Review by clarani — COMMENTED** [review]
Left some comments, I will finish my review tomorrow :)

**Comment by sim51** [comment]
FYI the bug on checkboxes has been fixed

File: front/src/common/BootstrapSNCF/SelectImprovedSNCF.tsx
Comment (clarani) [inline front/src/common/BootstrapSNCF/SelectImprovedSNCF.tsx]: Maybe this could be put in a const above
```
const displayAddOptionButton = withNewValueInpu && filterText.length > 0 && !filteredOptions.map((e) => (isObject(e) ? e.label : e)).includes(filterText);
```

File: front/src/common/Map/Settings/MapSettingsSpeedLimits.jsx
Comment (clarani) [inline front/src/common/Map/Settings/MapSettingsSpeedLimits.jsx:1]: You may have some conflicts on this file with Uriel's PR https://github.com/osrd-project/osrd/pull/4697

File: front/src/modules/scenario/components/AddOrEditScenarioModal.tsx
Comment (clarani) [inline front/src/modules/scenario/components/AddOrEditScenarioModal.tsx:259]: I think it's better to change the electricalProfilOptions initialization, line 63:
```
    ({
        ...response,
        electricalProfilOptions: [
          noElectricalProfileSetOption,
          ...sortBy(response.data, ['name']).map((option: ElectricalProfileSetType) => ({
            id: `${option.id}`,
            label: option.name,
          })),
        ],
      }),
```

and then 
```suggestion
                  options={electricalProfilOptions}
```

File: front/src/applications/operationalStudies/components/Study/AddOrEditStudyModal.tsx
Comment (clarani) [inline front/src/applications/operationalStudies/components/Study/AddOrEditStudyModal.tsx]: Why did you change that ? currentStudy is never null or undefined
```suggestion
                    value={{
                      id: currentStudy.study_type,
                      label: t(
                        `studyCategories.${currentStudy.study_type || 'nothingSelected'}`
                      ).toString(),
                    }}
```

File: front/src/applications/operationalStudies/components/Study/AddOrEditStudyModal.tsx
Comment (clarani) [inline front/src/applications/operationalStudies/components/Study/AddOrEditStudyModal.tsx]: Same here

```suggestion
                    value={{
                      id: currentStudy.state,
                      label: t(`studyStates.${currentStudy.state || 'nothingSelected'}`).toString(),
                    }}
```

File: front/src/applications/stdcm/components/OldAllowances/EmptyLine.tsx
Comment (clarani) [inline front/src/applications/stdcm/components/OldAllowances/EmptyLine.tsx:1]: This file is not used anymore I think, so I don't test it

**Review by clarani — COMMENTED** [review]
Tested everything in local, everything is working :)

File: front/src/common/BootstrapSNCF/SelectImprovedSNCF.tsx
Comment (clarani) [inline front/src/common/BootstrapSNCF/SelectImprovedSNCF.tsx]: Linter warning, you need to remove nextId import

File: front/src/common/BootstrapSNCF/SelectSNCF.tsx
Comment (clarani) [inline front/src/common/BootstrapSNCF/SelectSNCF.tsx:16]: Very nice, thanks :tada:

File: front/src/modules/scenario/components/AddOrEditScenarioModal.tsx
Comment (sim51) [reply front/src/modules/scenario/components/AddOrEditScenarioModal.tsx:259]: We have an other issue here : the select takes only string for ID, but in `electricalProfilOptions` it's a number. , so we need to cast it (on the onchange I do a `+e.id`). for the select component.

If I do it directly when we initialize it, we lost its primary type (same for the selectedValue)., which can leads to bad type if someone want to use it.

The type string is only needed for the select component, that's the cast is only done on it..

File: front/src/applications/operationalStudies/components/Study/AddOrEditStudyModal.tsx
Comment (sim51) [reply front/src/applications/operationalStudies/components/Study/AddOrEditStudyModal.tsx]: You've right ...
BTW, it's weird for me to have a default empty value instead of a placeholder. 
But the component doesn't accept placeholders. It can be an improvement for later

**Review by clarani — COMMENTED** [review]
LGTM, tested in local :heavy_check_mark:

**Review by clarani — APPROVED** [review]
Could you just squash your commits before merging ? Thanks !

**Pull request merged**
